File: signature-builder/python/dmarket_client.py

import json
from datetime import datetime
from urllib.parse import quote, urlencode

import requests
from nacl.bindings import crypto_sign


class DMarketClient:
    def __init__(self, public_key: str, secret_key: str):
        if not public_key or not secret_key:
            raise ValueError("Public and secret keys must be provided.")
        self._public_key = public_key
        self._secret_key = secret_key
        self._root_api_url = "https://api.dmarket.com"
        self._signature_prefix = "dmar ed25519 "

    def call(self, method: str, path: str, payload: dict = None):
        """
        Makes a signed API call to DMarket.

        :param method: HTTP method (e.g., 'GET', 'POST').
        :param path: API endpoint path (e.g., '/trade-aggregator/v1/last-sales').
        :param payload: Dictionary of parameters for the request.
        :return: A tuple of (response_json, error_string).
        """
        method = method.upper()
        nonce = str(round(datetime.now().timestamp()))
        api_url_path = path
        request_body = ""

        if payload:
            if method == "GET":
                api_url_path = f"{path}?{urlencode(payload, quote_via=quote)}"
            else:
                request_body = json.dumps(payload)

        string_to_sign = method + api_url_path + request_body + nonce
        signature = self._generate_signature(string_to_sign)

        headers = {
            "X-Api-Key": self._public_key,
            "X-Request-Sign": self._signature_prefix + signature,
            "X-Sign-Date": nonce,
        }
        if method not in ["GET"] and payload:
            headers["Content-Type"] = "application/json"

        full_url = self._root_api_url + api_url_path

        try:
            response = requests.request(
                method,
                full_url,
                headers=headers,
                data=request_body.encode('utf-8') if request_body else None
            )
            response.raise_for_status()
            return response.json(), None
        except requests.exceptions.RequestException as e:
            error_details = e.response.text if e.response else "No response body"
            return None, f"API call failed: {e}. Details: {error_details}"

    def _generate_signature(self, string_to_sign: str) -> str:
        encoded = string_to_sign.encode('utf-8')
        secret_bytes = bytes.fromhex(self._secret_key)
        signature_bytes = crypto_sign(encoded, secret_bytes)
        return signature_bytes[:64].hex()

File: signature-builder/js/dmarketClient.js
import https from 'https';
import nacl from 'tweetnacl';
import { URLSearchParams } from 'url';

// Helper functions from the original script
function byteToHexString(uint8arr) {
    if (!uint8arr) {
        return '';
    }
    let hexStr = '';
    for (let i = 0; i < uint8arr.length; i++) {
        let hex = (uint8arr[i] & 0xff).toString(16);
        hex = (hex.length === 1) ? '0' + hex : hex;
        hexStr += hex;
    }
    return hexStr;
}

function hexStringToByte(str) {
    if (typeof str !== 'string') {
        throw new TypeError('Wrong data type passed to convertor. Hexadecimal string is expected');
    }
    const uInt8arr = new Uint8Array(str.length / 2);
    for (let i = 0, j = 0; i < str.length; i += 2, j++) {
        uInt8arr[j] = parseInt(str.substr(i, 2), 16);
    }
    return uInt8arr;
}

export class DMarketClient {
    constructor(publicKey, secretKey) {
        if (!publicKey || !secretKey) {
            throw new Error('Public and secret keys must be provided.');
        }
        this.publicKey = publicKey;
        this.secretKey = secretKey;
        this.rootApiUrl = 'api.dmarket.com';
        this.signaturePrefix = 'dmar ed25519 ';
    }

    async call(method, path, payload = null) {
        method = method.toUpperCase();
        const timestamp = Math.floor(new Date().getTime() / 1000);
        let apiUrlPath = path;
        let requestBody = '';

        if (payload) {
            if (method === 'GET') {
                const params = new URLSearchParams(payload);
                apiUrlPath = `${path}?${params.toString()}`;
            } else {
                requestBody = JSON.stringify(payload);
            }
        }

        const stringToSign = method + apiUrlPath + requestBody + timestamp;
        const signature = this._generateSignature(stringToSign);

        const headers = {
            'X-Api-Key': this.publicKey,
            'X-Request-Sign': this.signaturePrefix + signature,
            'X-Sign-Date': timestamp,
        };

        if (method !== 'GET' && payload) {
            headers['Content-Type'] = 'application/json';
            headers['Content-Length'] = Buffer.byteLength(requestBody);
        }

        const options = {
            hostname: this.rootApiUrl,
            path: apiUrlPath,
            method: method,
            headers: headers,
        };

        return new Promise((resolve, reject) => {
            const req = https.request(options, (res) => {
                let data = '';
                res.on('data', (chunk) => {
                    data += chunk;
                });
                res.on('end', () => {
                    if (res.statusCode >= 400) {
                        reject(new Error(`API call failed with status code ${res.statusCode}: ${data}`));
                    } else {
                        try {
                            resolve(JSON.parse(data));
                        } catch (e) {
                            reject(new Error('Failed to parse JSON response.'));
                        }
                    }
                });
            });

            req.on('error', (e) => {
                reject(e);
            });

            if (method !== 'GET' && requestBody) {
                req.write(requestBody);
            }

            req.end();
        });
    }

    _generateSignature(stringToSign) {
        const secretKeyBytes = hexStringToByte(this.secretKey);
        // Use sign.detached for a signature format compatible with the other languages
        const signatureBytes = nacl.sign.detached(new TextEncoder('utf-8').encode(stringToSign), secretKeyBytes);
        return byteToHexString(signatureBytes);
    }
}
